### DISCUSSION TITLE
Validation errors from event hooks

### DISCUSSION THREAD

**Discussion by lyuboxa** [pocketbase/pocketbase#3464] [Q&A] [answered]
Hello, 

I'm trying to understand if validation errors can be surfaced to the user (ui/api). 
From my understanding a map of `validation.Error` should be safe to expose in the error response.

Unfortunately looking at the time when `ApiError` is created, any data passed that is not a map will be ignored. 
https://github.com/pocketbase/pocketbase/blob/7fa1ff53c9bc7c535a89d78ba491c18987e3bb48/apis/api_error.go#L77-L90

So if an event hook returns a single `validation.Error`, it will simply be ignored. Alternatively `validation.Errors` or `map[string]validation.Error` can be added to the switch to allow the passthrough and surface validation errors.

Can you recommend a way to expose validation errors through the records/collection crud hooks? Or is this just not the case right now.

Thanks!
Lyubo

**Comment by ganigeorgiev** [comment]
You can check https://pocketbase.io/docs/go-routing/#error-response.

**Reply by lyuboxa** [reply]
Thanks, this is the first place I looked but it seems like the validation error data does not get bubbled up to the API when raised by the a model update/create event, is this expected? 

See this tests

```golang
=== RUN   TestEventHook_Validataion
=== RUN   TestEventHook_Validataion/mode_update_validation_error
    api.go:114: Expected status code 422, got 400
    api.go:139: Cannot find "data":{"title":{"code":"invalid_title","message":"Invalid or missing title."}} in response body 
        {"code":400,"message":"Failed to update record.","data":{}}
=== RUN   TestEventHook_Validataion/record_update_validation_error
--- FAIL: TestEventHook_Validataion (0.03s)
    --- FAIL: TestEventHook_Validataion/mode_update_validation_error (0.02s)
    --- PASS: TestEventHook_Validataion/record_update_validation_error (0.01s)
```

```Golang
package test

import (
	"net/http"
	"strings"
	"testing"

	validation "github.com/go-ozzo/ozzo-validation/v4"
	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tests"
)

func TestEventHook_Validataion(t *testing.T) {
	scenarios := []tests.ApiScenario{
		{
			Name:   "model update validation error",
			Method: http.MethodPatch,
			Body:   strings.NewReader(`{"name":"test"}`),
			Url:    "/api/collections/users/records/4q1xlclmfloku33",
			RequestHeaders: map[string]string{
				// users, test@example.com
				"Authorization": "eyJhbGciOiJIUzI1NiJ9.eyJpZCI6IjRxMXhsY2xtZmxva3UzMyIsInR5cGUiOiJhdXRoUmVjb3JkIiwiY29sbGVjdGlvbklkIjoiX3BiX3VzZXJzX2F1dGhfIiwiZXhwIjoyMjA4OTg1MjYxfQ.UwD8JvkbQtXpymT09d7J6fdA0aP9g4FJ1GPh_ggEkzc",
			},
			ExpectedStatus:  422,
			ExpectedContent: []string{`"data":{"title":{"code":"invalid_title","message":"Invalid or missing title."}}`},
			ExpectedEvents: map[string]int{
				"OnModelBeforeUpdate":         1,
				"OnRecordBeforeUpdateRequest": 1,
			},
			TestAppFactory: testAppSetup,
			BeforeTestFunc: func(t *testing.T, app *tests.TestApp, _ *echo.Echo) {
				app.OnModelBeforeUpdate().Add(func(_ *core.ModelEvent) error {
					return apis.NewApiError(422, "something went wrong", map[string]validation.Error{
						"title": validation.NewError("invalid_title", "Invalid or missing title"),
					})
				})
			},
		},
		{
			Name:   "record update validation error",
			Method: http.MethodPatch,
			Body:   strings.NewReader(`{"name":"test"}`),
			Url:    "/api/collections/users/records/4q1xlclmfloku33",
			RequestHeaders: map[string]string{
				// users, test@example.com
				"Authorization": "eyJhbGciOiJIUzI1NiJ9.eyJpZCI6IjRxMXhsY2xtZmxva3UzMyIsInR5cGUiOiJhdXRoUmVjb3JkIiwiY29sbGVjdGlvbklkIjoiX3BiX3VzZXJzX2F1dGhfIiwiZXhwIjoyMjA4OTg1MjYxfQ.UwD8JvkbQtXpymT09d7J6fdA0aP9g4FJ1GPh_ggEkzc",
			},
			ExpectedStatus:  422,
			ExpectedContent: []string{`"data":{"title":{"code":"invalid_title","message":"Invalid or missing title."}}`},
			ExpectedEvents: map[string]int{
				"OnRecordBeforeUpdateRequest": 1,
			},
			TestAppFactory: testAppSetup,
			BeforeTestFunc: func(t *testing.T, app *tests.TestApp, _ *echo.Echo) {
				app.OnRecordBeforeUpdateRequest().Add(func(_ *core.RecordUpdateEvent) error {
					return apis.NewApiError(422, "something went wrong", map[string]validation.Error{
						"title": validation.NewError("invalid_title", "Invalid or missing title"),
					})
				})
			},
		},
	}

	for _, scenario := range scenarios {
		scenario.Test(t)
	}
}

func testAppSetup(t *testing.T) *tests.TestApp {
	t.Helper()

	testApp, err := tests.NewTestApp()
	if err != nil {
		t.Fatal(err)
	}

	return testApp
}
```

**Reply by ganigeorgiev** [reply, marked as answer]
>  but it seems like the validation error data does not get bubbled up to the API when raised by the a model update/create event, is this expected?

It is working as intended because the error returned from the db model events are always wrapped as part of the request API error, aka. internally when there is a form error you have something like:

```go
apis.NewBadRequestError("Failed to load the submitted data due to invalid formatting.", yourErr)
```

`ApiError` is usually not intended to be used as a generic error outside of the routing context (in your case the `*Request` hooks), but I'll consider eventually in the future to add checks and skip the extra wrapping in case `ApiError` is returned.

**Reply by lyuboxa** [reply]
Thanks for clarifying, I can see the double wrapping. Perhaps a mention of this in the documentation will be helpful.
From what I understand this can be mitigated when using custom routes, since there we can detect the bubbling. 

That said, it will be nice to be able to bubble up validation errors from all the event hooks. 

Thanks again.

**Reply by stamler** [reply]
@lyuboxa What did you end up doing? I'm at this stage in my project and trying to report cross-validation errors in a way that matches the native field validations (i.e. a JSON property for each field with a code and message property that I can surface in the UI)

**Reply by lyuboxa** [reply]
@stamler where I needed errors bubbled up, I started using custom apis.

**Reply by ganigeorgiev** [reply]
_Just for info in case someone stumble here, support for returning ApiError and validation errors from any hook was implemented as part of the [ongoing refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) and it will be available with the next v0.23.0 release soon._
